### PR TITLE
(Wait for #2071, #2075) [Memory] Add apply gradient step @open sesame 02/23 09:27

### PR DESCRIPTION
**Pull request by jihochu**
After derivative calculation, gradient needs to be applied. But, the
order for it was merged with derv calculation.
This patch divide apply graident order, and modify related lifespans.

Signed-off-by: Jiho Chu <jiho.chu@samsung.com>

### CODE REVIEW

**Review by DonghakPark — APPROVED** [review]
LGTM!

**Comment by SeoHyungjun** [comment]
LGTM

**Review by baek2sm — APPROVED** [review]
LGTM

File: nntrainer/tensor/manager.cpp [hunk 1]
```diff
@@ -371,8 +374,10 @@ std::vector<Weight *> Manager::requestWeights(
     auto grad_exec_order = default_grad_exec_order;
 
     if (trainable) {
-      var_exec_order.insert(var_exec_order.begin(), calcGradient_order);
-      grad_exec_order.insert(grad_exec_order.begin(), calcGradient_order);
+      var_exec_order.push_back(calcGradient_order);
+      var_exec_order.push_back(applyGradient_order);
```
Comment (jijoongmoon) [inline nntrainer/tensor/manager.cpp:378]: I don't think we need ```applyGradient_order``` for ```var_exec_order```. Could you check it once again?

File: nntrainer/tensor/manager.cpp [hunk 2]
```diff
@@ -498,7 +508,7 @@ Manager::requestInputs(const GraphNode &node,
   using RT = TensorSpecV2::RequestType;
 
   TensorSpecV2 var_common_spec, grad_common_spec;
-  var_common_spec.ls = TensorLifespan::FORWARD_GRAD_LIFESPAN;
+  var_common_spec.ls = TensorLifespan::FORWARD_GRAD_AGRAD_LIFESPAN;
```
Comment (jijoongmoon) [inline nntrainer/tensor/manager.cpp:511]: I think it should be FORWARD_GRAD_LIFESPAN. We do not need the Input Tensor for ```Applying Gradient```. Could you check plz?

File: nntrainer/tensor/manager.cpp [hunk 3]
```diff
@@ -351,12 +354,12 @@ void Manager::initializeTensorsTrain(unsigned int max_exec_order_) {
...
   const GraphNode &node, const std::vector<Weight::Spec> &weights_spec,
   bool trainable, const std::vector<std::string> &shared_names) {
-  const auto [forwarding_order, calcGradient_order, calcDerivative_order] =
+  const auto [forwarding_order, calcGradient_order, calcDerivative_order, applyGradient_order] =
     node.getExecutionOrder();
 
   std::vector<unsigned int> default_var_exec_order(
     {forwarding_order, calcDerivative_order});
-  std::vector<unsigned int> default_grad_exec_order({calcDerivative_order});
+  std::vector<unsigned int> default_grad_exec_order;
```
[the remark this answers is not in the document]
Comment (jihochu) [reply nntrainer/tensor/manager.cpp]: thanks. it removed.

File: nntrainer/tensor/manager.cpp (hunk 2 above)
Comment (jihochu) [reply nntrainer/tensor/manager.cpp]: It seems AGRAD is unnecessary. It passed test without it.
thianks.

**Review by jijoongmoon — APPROVED** [review]
LGTM

**Pull request merged**